### PR TITLE
 Add submariner orchestrator

### PR DESCRIPTION
**Pull request by tpantelis**
This is a new component/process that uses the Federator API to watch for cluster lifecycle events and create informers to distribute submariner Cluster and Endpoint resources across the cluster flotilla.

### CODE REVIEW

File: pkg/orchestrator/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	c.clusterWatcherMap[clusterID] = watcher
+
+	submarinerInformerFactory := submarinerInformers.NewSharedInformerFactoryWithOptions(submarinerClient, 0,
+		submarinerInformers.WithNamespace(submarinerNamespace))
+
+	clusterInformer := submarinerInformerFactory.Submariner().V1().Clusters().Informer()
+	clusterInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.clusterWorkQueue))
+
+	endpointInformer := submarinerInformerFactory.Submariner().V1().Endpoints().Informer()
+	endpointInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.endpointWorkQueue))
```
Comment (mpeterson) [inline pkg/orchestrator/controller/controller.go]: (unrelated to this line but the thought came here...)

There is a lot of quality code, good patterns and complex logic that is common to every watch that we need to set here. Something that would be repeated for most people doing the same (as here it is repeated for clusters and endpoints). 

I wonder if it would make sense to extrapolate these needs an provide a simpler interface for this?

File: pkg/orchestrator/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
+package controller
+
+import (
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: goimports does this.

File: pkg/orchestrator/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	endpointWorkQueue workqueue.RateLimitingInterface
+}
+
+type controller struct {
+	sync.Mutex
+	federator         federate.Federator
+	clusterWatcherMap map[string]*clusterWatcher
+
+	// Indirection hook for unit tests to supply fake client sets
+	clientSetFunc func(kubeConfig *rest.Config) (submarinerClientset.Interface, error)
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: We can rename it. I can't think of a better way with Go. This seems to be the pattern people use to hook in UTs. In Java I'd put the code in a protected method that could be overridden by a UT but you can't do it like that in Go.  Instead you can use indirection via function pointers.

File: pkg/orchestrator/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+)
+
+func init() {
+	flag.StringVar(&submarinerNamespace, "submariner-namespace", "submariner", "The namespace in which the submariner components are deployed.")
+}
+
+func New(federator federate.Federator) *controller {
+	return &controller{
+		federator:         federator,
+		clusterWatcherMap: make(map[string]*clusterWatcher),
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: gofmt does that and won't let me indent it.

File: pkg/orchestrator/controller/controller.go (hunk 1 above)
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: Let's open a JIRA to track it.

File: pkg/orchestrator/controller/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: removed the blank lines so goimports groups/sorts appropriately.

File: pkg/orchestrator/controller/controller.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+		return
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	var watcher *clusterWatcher
+	if watcher = c.clusterWatcherMap[clusterID]; watcher != nil {
+		klog.V(2).Infof("An entry for cluster %s already exists - stopping previous watch", clusterID)
+		watcher.close()
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: I guess we have to trust that the informer and workqueues do the right thing when told to close/shutdown.

File: pkg/orchestrator/controller/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+		mockFederator *mocks.MockFederator
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockFederator = mocks.NewMockFederator(mockCtrl)
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
```
Comment (mpeterson) [inline pkg/orchestrator/controller/controller_test.go]: nit: I'm a proponent of leaving the state as it was before the test. So here I'd nil both variables too.

(same for the other tests)

File: pkg/orchestrator/controller/controller.go [hunk 7]
```diff
@@ -59,6 +61,8 @@ func New(federator federate.Federator) *controller {
 }
 
 func (c *controller) Start() error {
+	submarinerv1.AddToScheme(scheme.Scheme)
```
Comment (mpeterson) [inline pkg/orchestrator/controller/controller.go]: this should be done in the main. The scheme is supposed to be populated/initialized at the beginning and not modified afterwards.

File: pkg/federate/kubefed/cluster.go [hunk 8]
```diff
@@ -153,7 +153,7 @@ func (f *federator) OnUpdate(oldObj, newObj interface{}) {
 	// (eg LastProbeTime). We're only interested in changes to the KubeFedClusterSpec which contains the
 	// cluster's connection info so check if the old and new Specs differ before proceeding.
 	if reflect.DeepEqual(oldCluster.Spec, newCluster.Spec) {
-		klog.V(2).Infof("KubeFedClusterSpecs are equal - not updating")
+		klog.V(3).Infof("KubeFedClusterSpecs are equal - not updating")
```
Comment (mpeterson) [inline pkg/federate/kubefed/cluster.go]: how is this change related?

File: pkg/orchestrator/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+	if err != nil {
+		return
+	}
+
+	labels := metadata.GetLabels()
+	if labels == nil {
+		return
+	}
+
+	label, exists = labels[key]
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: It made it simpler in this case but either way...

File: pkg/orchestrator/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+func newResourceEventHandler(clusterID string, workQueue workqueue.RateLimitingInterface) cache.ResourceEventHandler {
+	enqueue := func(obj interface{}, keyFunc func(obj interface{}) (string, error)) {
+		key, err := keyFunc(obj)
+		if err != nil {
+			klog.Errorf("Error creating MetaNamespaceKey: %v", err)
+			return
+		}
+
+		if label, exists := getLabel(obj, clusterIDLabelKey); exists && label != clusterID {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: sure either way

File: pkg/orchestrator/controller/controller.go [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+		workQueue.Add(key)
+	}
+
+	return &cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			enqueue(obj, cache.MetaNamespaceKeyFunc)
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			// TODO - handle updates or are the resources immutable?
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: Sure although I'm not sure of the significance of `user` here and haven't seen that used before.

File: pkg/orchestrator/controller/controller_test.go (hunk 6 above)
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: sure although it's not really necessary as each should be initialized in `BeforeEach`.

File: pkg/orchestrator/controller/controller.go (hunk 7 above)
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: This kind of is part of the beginning, also I thought it should the responsibility of this "class" to do it since it relies on it.  But in main is fine too - as long as it's done somewhere before the   informers start.

File: pkg/orchestrator/controller/controller_test.go (hunk 6 above)
Comment (mpeterson) [reply pkg/orchestrator/controller/controller_test.go]: I'm aware. Just stating my preference :)

File: pkg/orchestrator/controller/controller.go (hunk 7 above)
Comment (mpeterson) [reply pkg/orchestrator/controller/controller.go]: According to some "documentation" I've seen on the type registry, the preferred way is to add it to the registry before any interaction with the API and for that they usually recommend the main

File: pkg/orchestrator/controller/controller.go [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+		DeleteFunc: func(obj interface{}) {
+			enqueue(obj, cache.DeletionHandlingMetaNamespaceKeyFunc)
+		},
+	}
+}
+
+func (c *controller) OnAdd(clusterID string, kubeConfig *rest.Config) {
+	klog.V(2).Infof("In controller OnAdd for cluster %s", clusterID)
+
+	c.startNewWatcherForCluster(kubeConfig, clusterID)
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: I'm not sure I follow. Can you write the text that you would like to see?

[the remark this answers is not in the document]
Comment (mpeterson) [reply pkg/orchestrator/controller/controller.go]: Something along the lines of:

`// startNewWatcherForCluster locks the controller so there is no need to lock it here.`

To show why the usually expected locking for `OnAdd` and `OnDelete` is not there.

File: pkg/orchestrator/controller/controller.go [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+			return
+		}
+
+		klog.V(2).Infof("Processing key \"%s\" for cluster \"%s\"", key, clusterID)
+
+		func() {
+			defer workQueue.Done(key)
+
+			item, exists, err := informer.GetIndexer().GetByKey(key.(string))
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: yeah but I kind of preferred to use `if-elif-else` in this case since the conditions are related and it looks more natural to me. Also using the `return` approach actually adds more lines and I don't think it simplifies it.

File: pkg/federate/kubefed/cluster.go (hunk 8 above)
Comment (tpantelis) [reply pkg/federate/kubefed/cluster.go]: It's not technically but while running I noticed this is a bit chatty so I bumped the level.

File: pkg/orchestrator/controller/controller.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+		}
+
+		workQueue.Add(key)
+	}
+
+	return &cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			enqueue(obj, cache.MetaNamespaceKeyFunc)
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
```
Comment (mkolesnik) [inline pkg/orchestrator/controller/controller.go]: Since you're using ResourceEventHandlerFuncs you don't have to specify the UpdateFunc since it does nothign by default (the TODO can stay above the return).

File: pkg/orchestrator/controller/controller.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+	endpointInformer := submV1.Endpoints().Informer()
+	endpointInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.endpointWorkQueue))
+
+	submarinerInformerFactory.Start(watcher.stopChan)
+
+	go runQueueWorker(clusterID, c.federator, clusterInformer, watcher.clusterWorkQueue)
+	go runQueueWorker(clusterID, c.federator, endpointInformer, watcher.endpointWorkQueue)
+}
+
+func getLabel(obj interface{}, key string) (string, bool) {
```
Comment (mkolesnik) [inline pkg/orchestrator/controller/controller.go]: nit: currently it's only used with a constant so key can be removed as parameter

File: pkg/orchestrator/controller/controller.go [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+	// We can safely ignore the error as all the resources implement meta.Object
+	metadata, _ := meta.Accessor(copy)
+
+	labels := metadata.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels[key] = value
+	metadata.SetLabels(labels)
```
Comment (mkolesnik) [inline pkg/orchestrator/controller/controller.go]: Logically, souldn't this be done only when the labels is nil?

File: pkg/orchestrator/controller/controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mkolesnik) [reply pkg/orchestrator/controller/controller.go]: I actually tend to agree with the "return" method since it'll reduce the level of indentation which is already quite high (3 levels in for the normal flow), and make the code more readable since the normal flow is all on the same indentation level, while the exceptional cases are the ones that stray from the normal path.

File: pkg/orchestrator/controller/controller.go [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+func (c *controller) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, clusterWatcher := range c.clusterWatchers {
+		clusterWatcher.close()
+	}
+}
+
+func (c *controller) removeExistingClusterWatcher(clusterID string) {
```
Comment (mkolesnik) [inline pkg/orchestrator/controller/controller.go]: Perhaps it would be more correct to call this only n the context of update/remove since on add you don't expect it to actually exist already.

File: pkg/orchestrator/controller/controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller.go]: sure - I'm flexible either way - can refactor in followup.

**Comment by tpantelis** [comment]
> Some small comments, however I'm fine with merging as is for now and fixing in a later patch
> 
> I didn't have time to review the tests so if it's critical to merge now then do so and address the comments later

Comments addressed in https://github.com/submariner-io/admiral/pull/20

File: pkg/orchestrator/controller/controller_test.go [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+		endpoint         *submarinerv1.Endpoint
+		distributeCalled chan bool
+		mockCtrl         *gomock.Controller
+		mockFederator    *mocks.MockFederator
+		controller       *controller
+		fakeClientset    *fakeSubm.Clientset
+	)
+
+	BeforeEach(func() {
+		cluster = newCluster()
```
Comment (mkolesnik) [inline pkg/orchestrator/controller/controller_test.go]: I find it weird that you test clusters and endpoints in the same context, they could be each in it's own nested context.
This gets exacerbated when you add the delete and tests for that.

File: pkg/orchestrator/controller/controller_test.go [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+}
+
+type eqEndpoint struct {
+	expected *submarinerv1.Endpoint
+}
+
+var _ = Describe("Submariner Orchestrator", func() {
+	klog.InitFlags(nil)
+
+	When("start is called", testStart)
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: Stop was added later, mainly so the tests cleanup.

File: pkg/orchestrator/controller/controller_test.go [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+		It("should stop and restart the Submariner Endpoint and Cluster watches", func() {
+			testOnRemove("east")
+			watchReactor.reset()
+			testOnAdd("east")
+		})
+	})
+}
+
+func testResourceDistribution() {
+	const clusterID = "east"
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: sure

File: pkg/orchestrator/controller/controller_test.go [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+		}
+
+		controller.OnAdd(clusterID, &rest.Config{})
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+	})
+
+	createCluster := func() error {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: It is a helper func just not package scope to keep it localized.

File: pkg/orchestrator/controller/controller_test.go [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+	When("a Cluster resource is added with a non-matching clusterID label", func() {
+		It("should not distribute the resource", func() {
+			cluster.SetLabels(map[string]string{clusterIDLabelKey: "west"})
+			mockFederator.EXPECT().Distribute(EqCluster(cluster)).Return(nil).MaxTimes(0).Do(func(c *submarinerv1.Cluster) {
+				distributeCalled <- true
+			})
+
+			Expect(createCluster()).To(Succeed())
+
+			Consistently(distributeCalled).ShouldNot(Receive(), "Distribute was unexpectedly called")
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: Distribute is called async so we need to give it some time to ensure it's not called.

File: pkg/orchestrator/controller/controller_test.go [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+	w.endpointsWatchStarted = make(chan bool, 1)
+	w.clustersWatchStarted = make(chan bool, 1)
+}
+
+func (w *watchReactor) close() {
+	close(w.endpointsWatchStarted)
+	close(w.clustersWatchStarted)
+}
+
+func EqCluster(expected *submarinerv1.Cluster) *eqCluster {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: `EqCluster` would clash with the struct name.

File: pkg/orchestrator/controller/controller_test.go (hunk 18 above)
Comment (tpantelis) [reply pkg/orchestrator/controller/controller_test.go]: I plan to split them into separate functions.
